### PR TITLE
(PUP-8582) Add vendor_modules to basemodulepath

### PR DESCRIPTION
**Pull request by joshcooper**
Append the vendor_modules directory to the basemodulepath setting so that vendored modules are available in all environments.

On most platforms, the resulting basemodulepath corresponds to global, site and vendored modules directories, respectively.

On Windows, the resulting basemodulepath contains at least the global modules directory. Additionally if the env_windows_installdir fact[1] is set, which it is when running from puppet-agent packages, then we include the site and vendored modules directories.

For all platforms, the global modules directory ($codedir/modules) depends on whether we are running in a privileged context or not (so it's either in /etc/puppetlabs or ~/.puppetlabs).

[1] https://puppet.com/docs/facter/3.11/core_facts.html#envwindowsinstalldir

### CODE REVIEW

**Comment by joshcooper** [comment]
jenkins test this

**Comment by joshcooper** [comment]
jenkins please test this

**Comment by joshcooper** [comment]
Note this doesn't attempt to fix the precedence issues, just including vendored modules in the module search path

**Comment by joshcooper** [comment]
jenkins please test this

**Comment by jhelwig** [comment]
The logic is super-simple, but it might still be nice to have some tests for how the prefix logic works with Windows/not-Windows, and whether the Windows fact is present or not, to prevent accidental regressions? Unless you're thinking this is going to completely change once the precedence issues are addressed?

File: lib/puppet/defaults.rb [hunk 1]
```diff
@@ -1286,7 +1286,20 @@ def self.valid_file_checksum_types
       :desc       => "File that provides mapping between custom SSL oids and user-friendly names"
     },
     :basemodulepath => {
-      :default => "$codedir/modules#{File::PATH_SEPARATOR}/opt/puppetlabs/puppet/modules",
+      :default => lambda do
+        installdir =
+          if Puppet::Util::Platform.windows?
+            Facter.value(:env_windows_installdir) || 'C:/Program Files/Puppet Labs/Puppet'
```
Comment (melissa) [inline lib/puppet/defaults.rb]: This will work if a user decides to install the MSI under a directory that isn't `C:/Program Files`?

**Comment by melissa** [comment]
👍 to what Jacob said. I'm also curious where `$codedir` comes from.

```
$ be ./bin/puppet config print basemodulepath
/Users/melissa/.puppetlabs/etc/code/modules:/opt/puppetlabs/puppet/modules:/opt/puppetlabs/puppet/vendor_modules
```
but it's totes doing the right thing on my system. I do want to play with it a bit more, like to see what happens when different versions of a module are installed in different location, but from an initial once over, this looks like it should be good

**Comment by jhelwig** [comment]
@melissa `$codedir` should be coming from https://github.com/joshcooper/puppet/blob/077d162ff742be9bea41a709d13fbbb598f07a6b/lib/puppet/defaults.rb#L52-L60

**Comment by joshcooper** [comment]
jenkins please test this on windows2012r2-64,redhat7-64

File: spec/unit/defaults_spec.rb [hunk 2]
```diff
@@ -123,4 +123,32 @@
...
+
+    describe 'on windows', :if => Puppet::Util::Platform.windows? do
+      let(:installdir) { 'C:\Program Files\Puppet Labs\Puppet' }
+
+      it 'includes packaged modules if the env_windows_installdir fact is available' do
+        Facter.stubs(:value).with(:env_windows_installdir).returns(installdir)
+
+        expect(
+          Puppet[:basemodulepath]
+        ).to match(%r{.*/code/modules;#{installdir}/puppet/modules;#{installdir}/puppet/vendor_modules})
```
Comment (melissa) [inline spec/unit/defaults_spec.rb]: It looks like on windows we're only getting the first path `*/code/modules`, not all three as we would expect. Is it possible `Facter.value(:env_windows_installdir)` isn't returning anything for this test? If `installdir` is empty in the code, we aren't going to append `puppet/modules` or `puppet/vendor_modules` to the path
```diff
       -/.*\/code\/modules;C:\Program Files\Puppet Labs\Puppet\/puppet\/modules;C:\Program Files\Puppet Labs\Puppet\/puppet\/vendor_modules/
       +"C:/Users/appveyor/AppData/Local/Temp/1/rspecrun20180509-2900-k66xnz/tmp_settings20180509-2900-o944n5/code/modules"
```

**Comment by joshcooper** [comment]
jenkins please test this on windows2012r2-64,redhat7-64

**Comment by joshcooper** [comment]
jenkins please test this on windows2012r2-64a,redhat7-64a

**Comment by joshcooper** [comment]
jenkins please test this on windows2012r2-64a,redhat7-64a

**Comment by melissa** [comment]
jenkins please test this on windows2012r2-64a,redhat7-64a

File: lib/puppet/defaults.rb [hunk 3]
```diff
@@ -32,6 +32,20 @@ def self.valid_file_checksum_types
...
+    if Puppet::Util::Platform.windows?
+      path = ['$codedir/modules']
+      installdir = Facter.value(:env_windows_installdir)
+      if installdir
+        path << "#{installdir}/puppet/modules"
+        path << "#{installdir}/puppet/vendor_modules"
+      end
+      path.join(File::PATH_SEPARATOR)
+    else
+      '$codedir/modules:/opt/puppetlabs/puppet/modules:/opt/puppetlabs/puppet/vendor_modules'
```
Comment (tb3088) [inline lib/puppet/defaults.rb]: can we not define magic strings '/opt/puppetlabs/puppet' in the middle of a method and instead declare them as constants at the top? This should be settable by the user during compilation if they so choose during compilation.

**Pull request merged**